src/box/memtx_engine.cc

```diff
@@ -1371,6 +1371,7 @@ memtx_engine_memory_stat(struct engine *engine, struct engine_memory_stat *stat)
 
 static const struct engine_vtab memtx_engine_vtab = {
 	/* .free = */ memtx_engine_free,
+	/* .shutdown = */ NULL,
```

Comment:
You shouldn't set unimplemented methods to `NULL`. You should implement a generic version instead (maybe no-op), like we do for other methods.

Reply:
Done.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/blackhole.c b/src/box/blackhole.c
index 90e865623..805650428 100644
--- a/src/box/blackhole.c
+++ b/src/box/blackhole.c
@@ -174,7 +174,7 @@ blackhole_engine_create_space(struct engine *engine, struct space_def *def,
 
 static const struct engine_vtab blackhole_engine_vtab = {
 	/* .free = */ blackhole_engine_free,
-	/* .shutdown = */ NULL,
+	/* .shutdown = */ generic_engine_shutdown,
 	/* .create_space = */ blackhole_engine_create_space,
 	/* .create_read_view = */ generic_engine_create_read_view,
 	/* .prepare_join = */ generic_engine_prepare_join,
diff --git a/src/box/engine.c b/src/box/engine.c
index 7e6b70383..1348d428a 100644
--- a/src/box/engine.c
+++ b/src/box/engine.c
@@ -73,8 +73,7 @@ engine_shutdown(void)
 {
 	struct engine *engine, *tmp;
 	rlist_foreach_entry_safe(engine, &engines, link, tmp) {
-		if (engine->vtab->shutdown != NULL)
-			engine->vtab->shutdown(engine);
+		engine->vtab->shutdown(engine);
 	}
 }
 
@@ -477,4 +476,10 @@ generic_engine_check_space_def(struct space_def *def)
 	return 0;
 }
 
+void
+generic_engine_shutdown(struct engine *engine)
+{
+	(void)engine;
+}
+
 /* }}} */
diff --git a/src/box/engine.h b/src/box/engine.h
index 30ce6ebce..26db5f4b7 100644
--- a/src/box/engine.h
+++ b/src/box/engine.h
@@ -513,6 +513,7 @@ int generic_engine_backup(struct engine *, const struct vclock *,
 void generic_engine_memory_stat(struct engine *, struct engine_memory_stat *);
 void generic_engine_reset_stat(struct engine *);
 int generic_engine_check_space_def(struct space_def *);
+void generic_engine_shutdown(struct engine *engine);
 
 #if defined(__cplusplus)
 } /* extern "C" */
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 2cb4544f7..ad2781e8f 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -1371,7 +1371,7 @@ memtx_engine_memory_stat(struct engine *engine, struct engine_memory_stat *stat)
 
 static const struct engine_vtab memtx_engine_vtab = {
 	/* .free = */ memtx_engine_free,
-	/* .shutdown = */ NULL,
+	/* .shutdown = */ generic_engine_shutdown,
 	/* .create_space = */ memtx_engine_create_space,
 	/* .create_read_view = */ memtx_engine_create_read_view,
 	/* .prepare_join = */ memtx_engine_prepare_join,
diff --git a/src/box/service_engine.c b/src/box/service_engine.c
index bb0f81b4b..c60c8f1dd 100644
--- a/src/box/service_engine.c
+++ b/src/box/service_engine.c
@@ -89,7 +89,7 @@ service_engine_create_space(struct engine *engine, struct space_def *def,
 
 static const struct engine_vtab service_engine_vtab = {
 	/* .free = */ service_engine_free,
-	/* .shutdown = */ NULL,
+	/* .shutdown = */ generic_engine_shutdown,
 	/* .create_space = */ service_engine_create_space,
 	/* .create_read_view = */ generic_engine_create_read_view,
 	/* .prepare_join = */ generic_engine_prepare_join,
diff --git a/src/box/sysview.c b/src/box/sysview.c
index cc1d657d6..35cf947c6 100644
--- a/src/box/sysview.c
+++ b/src/box/sysview.c
@@ -558,7 +558,7 @@ sysview_engine_create_space(struct engine *engine, struct space_def *def,
 
 static const struct engine_vtab sysview_engine_vtab = {
 	/* .free = */ sysview_engine_free,
-	/* .shutdown = */ NULL,
+	/* .shutdown = */ generic_engine_shutdown,
 	/* .create_space = */ sysview_engine_create_space,
 	/* .create_read_view = */ generic_engine_create_read_view,
 	/* .prepare_join = */ generic_engine_prepare_join,
```

</details>